File: cakk-api/src/main/java/com/cakk/api/provider/oauth/impl/AppleAuthProvider.java
package com.cakk.api.provider.oauth.impl;

import java.security.PublicKey;

import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;

import com.cakk.api.provider.jwt.JwtProvider;
import com.cakk.api.provider.oauth.OidcProvider;
import com.cakk.api.provider.oauth.PublicKeyProvider;
import com.cakk.client.vo.OidcPublicKeyList;
import com.cakk.client.web.AppleAuthClient;

@Component
@RequiredArgsConstructor
public class AppleAuthProvider implements OidcProvider {

	private final AppleAuthClient appleAuthClient;
	private final PublicKeyProvider publicKeyProvider;
	private final JwtProvider jwtProvider;

	@Override
	public String getProviderId(final String idToken) {
		final OidcPublicKeyList oidcPublicKeyList = appleAuthClient.getPublicKeys();
		final PublicKey publicKey = publicKeyProvider.generatePublicKey(parseHeaders(idToken), oidcPublicKeyList);

		return jwtProvider.parseClaims(idToken, publicKey).getSubject();
	}
}

File: cakk-api/src/main/kotlin/com/cakk/api/provider/oauth/AppleAuthProvider.kt
package com.cakk.api.provider.oauth

import org.springframework.stereotype.Component

import com.cakk.api.provider.jwt.JwtProviderImpl
import com.cakk.core.provider.oauth.OidcProvider
import com.cakk.external.client.AppleAuthClient

@Component
class AppleAuthProvider(
	private val appleAuthClient: AppleAuthClient,
	private val publicKeyProvider: PublicKeyProvider,
	private val jwtProviderImpl: JwtProviderImpl
) : OidcProvider {

	override fun getProviderId(idToken: String): String {
		val oidcPublicKeyList = appleAuthClient.getPublicKeys()
		val publicKey = publicKeyProvider.generatePublicKey(parseHeaders(idToken), oidcPublicKeyList)

		return jwtProviderImpl.parseClaims(idToken, publicKey).subject
	}
}
